server.go

```diff
@@ -0,0 +1,448 @@
+		debugStream:   debugStream,
+		debugLevel:    debugLevel,
+		readOnly:      readOnly,
+		rootDir:       rootDir,
+		pktChan:       make(chan rxPacket, workerCount),
+		openFiles:     map[string]*os.File{},
+		openFilesLock: &sync.RWMutex{},
+		maxTxPacket:   1 << 15,
+		WorkerCount:   workerCount,
+	}, nil
+}
+
+type rxPacket struct {
+	pktType  fxp
+	pktBytes []byte
+}
+
+// Unmarshal a single logical packet from the secure channel
+func (svr *Server) rxPackets() error {
+	defer close(svr.pktChan)
+
+	for {
+		pktType, pktBytes, err := recvPacket(svr.in)
+		if err == io.EOF {
+			fmt.Fprintf(svr.debugStream, "rxPackets loop done\n")
+			return nil
+		} else if err != nil {
+			fmt.Fprintf(svr.debugStream, "recvPacket error: %v\n", err)
+			return err
+		}
+
+		svr.pktChan <- rxPacket{fxp(pktType), pktBytes}
+	}
+}
+
+// Up to N parallel servers
+func (svr *Server) sftpServerWorker(doneChan chan error) {
+	for pkt := range svr.pktChan {
+		if pkt, err := svr.decodePacket(pkt.pktType, pkt.pktBytes); err != nil {
+			fmt.Fprintf(svr.debugStream, "decodePacket error: %v\n", err)
+			doneChan <- err
+			return
+		} else {
+			//fmt.Fprintf(svr.debugStream, "pkt: %T %v\n", pkt, pkt)
+			pkt.respond(svr)
+		}
+	}
+	doneChan <- nil
+}
+
+// Run this server until the streams stop or until the subsystem is stopped
+func (svr *Server) Run() error {
+	if svr.WorkerCount <= 0 {
+		return fmt.Errorf("sftp server requires > 0 workers")
+	}
+	go svr.rxPackets()
+	doneChan := make(chan error)
+	for i := 0; i < svr.WorkerCount; i++ {
+		go svr.sftpServerWorker(doneChan)
+	}
+	for i := 0; i < svr.WorkerCount; i++ {
+		if err := <-doneChan; err != nil {
+			// abort early and shut down the session on un-decodable packets
+			break
+		}
+	}
+	fmt.Fprintf(svr.debugStream, "sftp server run finished\n")
+	return svr.out.Close()
+}
+
+func (svr *Server) decodePacket(pktType fxp, pktBytes []byte) (serverRespondablePacket, error) {
+	//pktId, restBytes := unmarshalUint32(pktBytes[1:])
+	var pkt serverRespondablePacket = nil
+	switch pktType {
+	case ssh_FXP_INIT:
+		pkt = &sshFxInitPacket{}
+	case ssh_FXP_LSTAT:
+		pkt = &sshFxpLstatPacket{}
+	case ssh_FXP_VERSION:
+	case ssh_FXP_OPEN:
+		pkt = &sshFxpOpenPacket{}
+	case ssh_FXP_CLOSE:
+		pkt = &sshFxpClosePacket{}
+	case ssh_FXP_READ:
+		pkt = &sshFxpReadPacket{}
+	case ssh_FXP_WRITE:
+		pkt = &sshFxpWritePacket{}
+	case ssh_FXP_FSTAT:
+		pkt = &sshFxpFstatPacket{}
+	case ssh_FXP_SETSTAT:
+	case ssh_FXP_FSETSTAT:
+	case ssh_FXP_OPENDIR:
+		pkt = &sshFxpOpendirPacket{}
+	case ssh_FXP_READDIR:
+		pkt = &sshFxpReaddirPacket{}
+	case ssh_FXP_REMOVE:
+		pkt = &sshFxpRemovePacket{}
+	case ssh_FXP_MKDIR:
+		pkt = &sshFxpMkdirPacket{}
+	case ssh_FXP_RMDIR:
+	case ssh_FXP_REALPATH:
+		pkt = &sshFxpRealpathPacket{}
+	case ssh_FXP_STAT:
+		pkt = &sshFxpStatPacket{}
+	case ssh_FXP_RENAME:
+		pkt = &sshFxpRenamePacket{}
+	case ssh_FXP_READLINK:
+		pkt = &sshFxpReadlinkPacket{}
+	case ssh_FXP_SYMLINK:
```

Comment:
if these are cases that should be ignored, then do

case ssh_FXP_STATUS, ssh_FXP_HANDLE, ...:
   // continue